### PR TITLE
Preprocessor shebang clipping fix

### PR DESCRIPTION
**Pull request by jasonm23**
RFC @shayne @kasper

When we capture the output from babel, coffee etc. it _should_ no longer include the shebang, so we should not read past that position.

Perhaps some preprocessors might leave in things like that? Please let me know if they do.

It would be a simple solve to just read past it, if it's there, but not blindly.

### CODE REVIEW

**Comment by jasonm23** [comment]
If the shebang is included we can do this optionally... 
```
    // Read past shebang-directive
    NSString *outputHead = [standardOutputFile readDataOfLength:scanner.scanLocation];
    if (![command isEqualToString:outputHead]) {
      // reset seek if the output doesn't include the shebang-directive
      [standardOutputFile seekToFileOffset:0]
    }
```

**Comment by jasonm23** [comment]
Since we can't rewind the std out file handler, we can just re-assemble the output instead....

    // Read past shebang-directive
    NSString *outputHead = [[NSString alloc] initWithData:[standardOutputFile readDataOfLength:scanner.scanLocation]
                                                 encoding:NSUTF8StringEncoding];
    NSString *processed = [[NSString alloc] initWithData:[standardOutputFile readDataToEndOfFile] encoding:NSUTF8StringEncoding];

    // return the output with shebang clipped
    if ([command isEqualToString:outputHead]) {

      return processed;

    }

    // otherwise use the full output of the preprocessor
    return [outputHead stringByAppendingString:processed];

**Comment by jasonm23** [comment]
Seems to work as expected

**Comment by jasonm23** [comment]
Ok, this should be ready for a merge. @kasper Let me know when you can :+1:

**Comment by jasonm23** [comment]
Tested with a complicated shebang.

```bash
#!PATH=/usr/local/bin:$PATH /usr/bin/env coffee -c -l -p -b
```

**Comment by kasper** [comment]
@jasonm23 Good catch! I was wondering about this, it would seem that some indeed leave it in the output. @shayne Could you confirm this?

File: Phoenix/PHShebangPreprocessor.m [hunk 1]
```diff
@@ -74,10 +74,24 @@ + (NSString *) process:(NSString *)script atPath:(NSString *)path error:(NSError
         return script;
     }
 
-    // Read past shebang-directive
-    [standardOutputFile readDataOfLength:scanner.scanLocation];
+    // Read past shebang-directive (hold onto it incase it's not the shebang text)
```
Comment (kasper) [inline Phoenix/PHShebangPreprocessor.m]: I’m wondering if it might be easier and safer to create a new scanner for the output. Check if the command is present with `readCommand:`, if it is, the scanner will be moved forward, if not it, will stay intact. Then just return the output string based on the position of the scanner. (I’m thinking of an edge case where the preprocessor might even reformat the shebang.)

**Comment by kasper** [comment]
@jasonm23 What is the most minimal shebang to test this against CoffeeScript?

**Comment by jasonm23** [comment]
Since we use `/bin/bash` without a login we need to use 

   #!/usr/local/bin/coffee -bcp

For regular coffeescript, or 

   #!/usr/local/bin/coffee -bcpl

For literate coffeescript.

These would assume that `npm install -g coffee-script` completed successfully

**Comment by jasonm23** [comment]
We should probably have tests for these:

- babel
- coffeescript
- typescript
- livescript
- clojurescript

any others?

File: Phoenix/PHShebangPreprocessor.m (hunk 1 above)
Comment (jasonm23) [inline Phoenix/PHShebangPreprocessor.m]: Didn't think it was really worth running another scan, when we can easily catch the shebang length substring, and join them again if required.

We could also just grab `availableData`

**Comment by kasper** [comment]
@jasonm23 Thanks! I was a bit lazy with testing this manually only on a few commands.

(Unfortunately, this project didn’t have test to begin with, so I had to make the hard decision to save my time and evolve the project without them.

Now that the project will stabilise, I would be more than happy to start adding tests to the project. Unfortunately, Travis has an outstanding issue #4473 which seems to prevent running integration tests against the Accessibility API. Unit tests should be fine though.)

File: Phoenix/PHShebangPreprocessor.m (hunk 1 above)
Comment (kasper) [inline Phoenix/PHShebangPreprocessor.m]: What if the user has a weirdly formatted shebang, or leading white spaces and/or new lines, and the preprocessor reformats this? Then the command will not match against the output.

**Comment by jasonm23** [comment]
I'm just doing side testing of babel now (with ES2015 preset)

**Comment by jasonm23** [comment]
By the way am I correct in thinking we can't reset the seek offset for a StandardInput file handle (or any other pipe as file handle!)

**Comment by jasonm23** [comment]
- babel leaves in the shebang.
  -[ ] yes
  -[ ] no

- coffeescript leaves in the shebang.
  -[ ] yes
  -[ ] no

- typescript leaves in the shebang.
  -[ ] yes
  -[ ] no

- livescript leaves in the shebang.
  -[ ] yes
  -[ ] no

- clojurescript leaves in the shebang.
  -[ ] yes
  -[ ] no

**Comment by kasper** [comment]
@jasonm23 Wouldn’t https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSFileHandle_Class/#//apple_ref/occ/instm/NSFileHandle/seekToFileOffset: work?

Thanks for doing the manual testing! I appreciate it. Could you also test, if you add a few newlines before the shebang, are they removed from the output. Babel’s online interpreter seems to remove them. I’m currently mobile and will be able to resolve this later today.

**Comment by jasonm23** [comment]
> if you add a few newlines before the shebang

Would that still be a legal shebang? IIRC it's not.

**Comment by jasonm23** [comment]
`seekToFileOffset` did not seem to work when I tried it.  I'll give it another try now.

**Comment by jasonm23** [comment]
By the way, the most minimal shebang for any node based preprocessor, will probably need to include the fix for `PATH=/usr/local/bin:$PATH` (or it won't find node)

so: 

    #! PATH=/usr/local/bin:$PATH /usr/local/bin/coffee -bcp

    #! PATH=/usr/local/bin:$PATH /usr/local/bin/coffee -bcpl 

    #! PATH=/usr/local/bin:$PATH /usr/local/bin/tsc (not sure the options for typescript)

    #! PATH=/usr/local/bin:$PATH /usr/local/bin/lsc -bcp 

    #! PATH=/usr/local/bin:$PATH /usr/local/bin/cljsc ... (I don't know how to get the JS sent to stdout from cljsc at this time, only to send to external files)

**Comment by jasonm23** [comment]
BTW: Babel is "fun" in that the presets need to be installed in the home folder `/node_modules` (they don't seem to be found when using `npm install -g ...`) also `~/.babelrc` must include the preset(s) to be used.

@shayne I assume you're using babel, do you think we need some helper notes (or linked resource?)

**Comment by kasper** [comment]
@jasonm23 Yeah, I’m not sure if it’s literally a legal shebang (the specification doesn’t really state if whitespace characters are allowed before the directive), but NSScanner will skip any newlines and whitespaces, so it will become one nonetheless. This is why, I assume, in an edge case the string comparison will fail.

**Comment by jasonm23** [comment]
I see, well I think it's fair to specify correct usage in documentation.

Put that edge case down to use error

**Comment by kasper** [comment]
@jasonm23 Thanks for your time looking at this issue! It’s now merged. I simplified the logic a bit and it should now be fixed.

**Comment by kasper** [comment]
@jasonm23 The user’s PATH is now read, so the most minimal CoffeeScript setup seems to be `#!/usr/bin/env coffee -p`.

@shayne Jason is probably right that if we mention Babel in the API, we should probably have notes or a linked resource if the setup requires any additional steps other than a basic install.

**Comment by jasonm23** [comment]
We can apply the same rule to all pre processors to keep it simple.

E.g. Shebangs can only be specified on the initial line of the config script. 

@kasper Also note that if the user has a shell other than bash as their default, reading path via bash --login won't assist them much.

A lot of people are using zsh and a significant minority use Fish.  I expect there's a few using other shells too.  We should probably try to discover the user shell and run that?

**Comment by shayne** [comment]
Thanks for digging into this more, guys and apologies for the belated response.

Babel just upgraded to a 6.0 release which shifted all built-in functionality into its plugin architecture which is what's causing the kludge of issues @jasonm23 describes. When I started working on my Phoenix config 5.8 as out and it is what I'm using still. I'll probably upgrade to 6.0 soon, if anything to iron out what @jasonm23 outlined above.

For now, I'd be fine if Babel was removed as an example. If I can come up with a reasonable minimal Babel setup I'll open a PR.

**Comment by jasonm23** [comment]
Hi @shayne thanks, I think it's perfectly good to strip the shebang when it's there.  As you can see above Typescript also does this, and we should really allow for the use of any preprocessor.

Some are going to generate completely new output, some will leave alone things they don't recognise.  It makes a _sort of sense_ (even if they are invalid in the target language/dialect.)

Given it's popularity, I don't see there's any need to remove Babel from the examples.

**Comment by jasonm23** [comment]
@kasper we should take steps to use the current user's shell instead of hardcoding `/bin/bash`.

Maybe we can use something like [getpwuid](https://developer.apple.com/library/ios/documentation/System/Conceptual/ManPages_iPhoneOS/man3/getpwuid.3.html)

For example:

    #include <pwd.h>

    char get_shell(uid) {
      struct passwd *pw = getpwuid(uid);
      return pw.pw_shell;
    }

@shayne @kasper do you know of a more Cocoa/NS canonical way to do this?

**Comment by jasonm23** [comment]
This example seems to work, needs cleaning up.

```c
#include <pwd.h>
register uid_t uid;
struct passwd *uid_pw;
uid = geteuid ();
uid_pw = getpwuid (uid);
strcpy(uname,uid_pw->pw_shell)
```

**Comment by jasonm23** [comment]
Ahh ...

```objc
NSDictionary *environmentDict = [[NSProcessInfo processInfo] environment];
NSString *shellString = [environmentDict objectForKey:@"SHELL"];
```

This looks better.

**Comment by jasonm23** [comment]
#49

**Comment by kasper** [comment]
@jasonm23 Yes, Phoenix expects to see the shebang as the first directive in the file. Although this is standard, I’ll make a note about in the API.

@shayne I’m happy to keep Babel as an example, since it’s completely reasonable that someone will want to use the new syntax. So, if you have time, you could write for instance a Wiki-page which we could link to the API.

We should now have preprocessor support for basically anything. You could even write your own shell script. :)
